File: 2020/day06.py
from collections import Counter
from itertools import chain

with open("inputs/day06.input", "r") as f:
    data = f.read()
    groups = data.strip().split("\n\n")


def part1(groups):
    return sum(len(Counter(chain(*(a for a in group.split("\n")))))
               for group in groups)


def part2(groups):
    return sum(sum(a == len(group.split("\n"))
                   for a in Counter(chain(*(a for a in group.split("\n"))))
                   .values())
               for group in groups)


print(f"Part 1: {part1(groups)}")
print(f"Part 2: {part2(groups)}")

File: rust-2020/src/day06.rs
pub fn read(contents: &str) -> (u32, u32) {
    let mut a = 0_u32;
    let mut b = 0_u32;
    for group in contents.trim().split("\n\n") {
        a += group
            .bytes()
            .filter(|&c| c != b'\n')
            .fold(0_u32, |acc, choice| acc | 1 << (choice - b'a'))
            .count_ones();
        b += group
            .split('\n')
            .map(|person| {
                person
                    .bytes()
                    .fold(0_u32, |acc, choice| acc | 1 << (choice - b'a'))
            })
            .fold(std::u32::MAX, |acc, one| acc & one)
            .count_ones();
    }
    return (a, b);
}

pub fn part1(data: &(u32, u32)) -> u32 {
    data.0
}
pub fn part2(data: &(u32, u32)) -> u32 {
    data.1
}
pub fn run() -> (String, String) {
    let contents = include_str!(concat!(env!("CARGO_MANIFEST_DIR"), "/inputs/day06.input"));
    let data = read(&contents);
    let p1 = format!("{}", part1(&data));
    let p2 = format!("{}", part2(&data));
    return (p1, p2);
}
